File: services/notebook-runtime-service/src/domain/environment.rs
use std::path::{Component, Path, PathBuf};

use chrono::{DateTime, Utc};
use tokio::fs;
use uuid::Uuid;

use crate::models::workspace::NotebookWorkspaceFile;

pub fn notebook_workspace_root(data_dir: &str, notebook_id: Uuid) -> PathBuf {
    Path::new(data_dir)
        .join("workspaces")
        .join(notebook_id.to_string())
}

pub async fn ensure_workspace_seed(data_dir: &str, notebook_id: Uuid) -> Result<(), String> {
    let root = notebook_workspace_root(data_dir, notebook_id);
    fs::create_dir_all(&root)
        .await
        .map_err(|error| format!("failed to create notebook workspace: {error}"))?;

    let readme = root.join("README.md");
    if fs::metadata(&readme).await.is_err() {
        let seed = format!(
            "# Notebook Workspace\n\nNotebook `{}` now has a persisted workspace.\n\nUse this area for helper scripts, prompts, notes, and analysis artifacts that live next to your notebook cells.\n",
            notebook_id
        );
        fs::write(&readme, seed)
            .await
            .map_err(|error| format!("failed to seed notebook workspace: {error}"))?;
    }

    Ok(())
}

pub async fn list_workspace_files(
    data_dir: &str,
    notebook_id: Uuid,
) -> Result<Vec<NotebookWorkspaceFile>, String> {
    ensure_workspace_seed(data_dir, notebook_id).await?;
    let root = notebook_workspace_root(data_dir, notebook_id);
    let mut files = Vec::new();
    collect_workspace_files(&root, &root, &mut files).await?;
    files.sort_by(|left, right| left.path.cmp(&right.path));
    Ok(files)
}

pub async fn upsert_workspace_file(
    data_dir: &str,
    notebook_id: Uuid,
    path: &str,
    content: &str,
) -> Result<NotebookWorkspaceFile, String> {
    ensure_workspace_seed(data_dir, notebook_id).await?;
    let normalized = normalize_workspace_path(path)?;
    let root = notebook_workspace_root(data_dir, notebook_id);
    let absolute_path = root.join(&normalized);

    if let Some(parent) = absolute_path.parent() {
        fs::create_dir_all(parent)
            .await
            .map_err(|error| format!("failed to create workspace directories: {error}"))?;
    }

    fs::write(&absolute_path, content)
        .await
        .map_err(|error| format!("failed to write workspace file: {error}"))?;

    load_workspace_file(&root, &absolute_path).await
}

pub async fn delete_workspace_file(
    data_dir: &str,
    notebook_id: Uuid,
    path: &str,
) -> Result<bool, String> {
    ensure_workspace_seed(data_dir, notebook_id).await?;
    let normalized = normalize_workspace_path(path)?;
    let root = notebook_workspace_root(data_dir, notebook_id);
    let absolute_path = root.join(normalized);

    match fs::metadata(&absolute_path).await {
        Ok(metadata) if metadata.is_file() => {
            fs::remove_file(&absolute_path)
                .await
                .map_err(|error| format!("failed to delete workspace file: {error}"))?;
            Ok(true)
        }
        Ok(_) => Err("workspace path is not a file".to_string()),
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(false),
        Err(error) => Err(format!("failed to inspect workspace file: {error}")),
    }
}

fn normalize_workspace_path(path: &str) -> Result<PathBuf, String> {
    let candidate = path.trim().replace('\\', "/");
    if candidate.is_empty() {
        return Err("workspace path is required".to_string());
    }

    let raw = Path::new(&candidate);
    if raw.is_absolute() {
        return Err("workspace paths must be relative".to_string());
    }

    let mut normalized = PathBuf::new();
    for component in raw.components() {
        match component {
            Component::Normal(part) => normalized.push(part),
            Component::CurDir => {}
            Component::ParentDir => {
                return Err("workspace paths cannot escape the notebook root".to_string());
            }
            Component::RootDir | Component::Prefix(_) => {
                return Err("workspace paths must stay inside the notebook root".to_string());
            }
        }
    }

    if normalized.as_os_str().is_empty() {
        return Err("workspace path is required".to_string());
    }

    Ok(normalized)
}

async fn collect_workspace_files(
    root: &Path,
    current: &Path,
    files: &mut Vec<NotebookWorkspaceFile>,
) -> Result<(), String> {
    let mut entries = fs::read_dir(current)
        .await
        .map_err(|error| format!("failed to read workspace directory: {error}"))?;

    while let Some(entry) = entries
        .next_entry()
        .await
        .map_err(|error| format!("failed to iterate workspace directory: {error}"))?
    {
        let path = entry.path();
        let metadata = entry
            .metadata()
            .await
            .map_err(|error| format!("failed to stat workspace entry: {error}"))?;

        if metadata.is_dir() {
            Box::pin(collect_workspace_files(root, &path, files)).await?;
            continue;
        }

        files.push(load_workspace_file(root, &path).await?);
    }

    Ok(())
}

async fn load_workspace_file(
    root: &Path,
    absolute_path: &Path,
) -> Result<NotebookWorkspaceFile, String> {
    let content = fs::read_to_string(absolute_path)
        .await
        .map_err(|error| format!("failed to read workspace file: {error}"))?;
    let metadata = fs::metadata(absolute_path)
        .await
        .map_err(|error| format!("failed to stat workspace file: {error}"))?;
    let relative = absolute_path
        .strip_prefix(root)
        .map_err(|error| format!("failed to compute workspace relative path: {error}"))?;
    let modified = metadata
        .modified()
        .ok()
        .map(DateTime::<Utc>::from)
        .unwrap_or_else(Utc::now);

    Ok(NotebookWorkspaceFile {
        path: relative.to_string_lossy().replace('\\', "/"),
        language: infer_workspace_language(relative),
        content,
        size_bytes: metadata.len() as usize,
        updated_at: modified,
    })
}

fn infer_workspace_language(path: &Path) -> String {
    match path
        .extension()
        .and_then(|extension| extension.to_str())
        .map(|value| value.to_ascii_lowercase())
    {
        Some(extension) if extension == "py" => "python".to_string(),
        Some(extension) if extension == "sql" => "sql".to_string(),
        Some(extension) if extension == "r" => "r".to_string(),
        Some(extension) if extension == "md" => "markdown".to_string(),
        Some(extension) if extension == "json" => "json".to_string(),
        Some(extension) if extension == "ts" || extension == "tsx" => "typescript".to_string(),
        Some(extension) if extension == "js" || extension == "jsx" => "javascript".to_string(),
        Some(extension) if extension == "toml" => "toml".to_string(),
        _ => "text".to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_parent_directory_escapes() {
        let error = normalize_workspace_path("../secret.txt").expect_err("path should fail");
        assert!(error.contains("escape"));
    }

    #[test]
    fn normalizes_relative_workspace_paths() {
        let path = normalize_workspace_path("src/./analysis.py").expect("path should normalize");
        assert_eq!(path.to_string_lossy(), "src/analysis.py");
    }
}

File: services/notebook-runtime-service/internal/domain/environment/environment.go
// Package environment ports services/notebook-runtime-service/src/domain/environment.rs
// 1:1: workspace seeding + path safety helpers used by the notebook
// kernels (each notebook gets a sandboxed directory under DATA_DIR).
//
// The path normaliser is the security-relevant piece: it must reject
// `..`, absolute paths and Windows-style drive prefixes so a kernel
// cell can never write outside its notebook's workspace root.
package environment

import (
	"errors"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/google/uuid"

	"github.com/openfoundry/openfoundry-go/services/notebook-runtime-service/internal/models"
)

// WorkspaceRoot returns the per-notebook workspace path under dataDir.
// Mirrors `notebook_workspace_root` in Rust.
func WorkspaceRoot(dataDir string, notebookID uuid.UUID) string {
	return filepath.Join(dataDir, "workspaces", notebookID.String())
}

// EnsureSeed makes sure the workspace exists and contains the seed
// README so a freshly-created notebook has a non-empty file tree the
// kernel can navigate.
func EnsureSeed(dataDir string, notebookID uuid.UUID) error {
	root := WorkspaceRoot(dataDir, notebookID)
	if err := os.MkdirAll(root, 0o755); err != nil {
		return err
	}
	readme := filepath.Join(root, "README.md")
	if _, err := os.Stat(readme); errors.Is(err, os.ErrNotExist) {
		seed := "# Notebook Workspace\n\nNotebook `" + notebookID.String() +
			"` now has a persisted workspace.\n\nUse this area for helper scripts, prompts, notes, " +
			"and analysis artifacts that live next to your notebook cells.\n"
		if err := os.WriteFile(readme, []byte(seed), 0o644); err != nil {
			return err
		}
	}
	return nil
}

// ListWorkspaceFiles enumerates all files under the workspace, sorted
// by path for deterministic ordering (matches Rust).
func ListWorkspaceFiles(dataDir string, notebookID uuid.UUID) ([]models.NotebookWorkspaceFile, error) {
	if err := EnsureSeed(dataDir, notebookID); err != nil {
		return nil, err
	}
	root := WorkspaceRoot(dataDir, notebookID)
	var out []models.NotebookWorkspaceFile

	walkErr := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			return nil
		}
		f, err := loadFile(root, path, info)
		if err != nil {
			return err
		}
		out = append(out, f)
		return nil
	})
	if walkErr != nil {
		return nil, walkErr
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Path < out[j].Path })
	return out, nil
}

// UpsertWorkspaceFile writes (or overwrites) a single file in the
// workspace. The path must be relative and stay inside the root.
func UpsertWorkspaceFile(dataDir string, notebookID uuid.UUID, relPath, content string) (models.NotebookWorkspaceFile, error) {
	if err := EnsureSeed(dataDir, notebookID); err != nil {
		return models.NotebookWorkspaceFile{}, err
	}
	clean, err := NormalizePath(relPath)
	if err != nil {
		return models.NotebookWorkspaceFile{}, err
	}
	root := WorkspaceRoot(dataDir, notebookID)
	abs := filepath.Join(root, clean)
	if err := os.MkdirAll(filepath.Dir(abs), 0o755); err != nil {
		return models.NotebookWorkspaceFile{}, err
	}
	if err := os.WriteFile(abs, []byte(content), 0o644); err != nil {
		return models.NotebookWorkspaceFile{}, err
	}
	info, err := os.Stat(abs)
	if err != nil {
		return models.NotebookWorkspaceFile{}, err
	}
	return loadFile(root, abs, info)
}

// DeleteWorkspaceFile removes a file under the workspace. Returns
// (false, nil) when the file did not exist; true on success.
func DeleteWorkspaceFile(dataDir string, notebookID uuid.UUID, relPath string) (bool, error) {
	if err := EnsureSeed(dataDir, notebookID); err != nil {
		return false, err
	}
	clean, err := NormalizePath(relPath)
	if err != nil {
		return false, err
	}
	abs := filepath.Join(WorkspaceRoot(dataDir, notebookID), clean)
	info, err := os.Stat(abs)
	if errors.Is(err, os.ErrNotExist) {
		return false, nil
	}
	if err != nil {
		return false, err
	}
	if !info.Mode().IsRegular() {
		return false, errors.New("workspace path is not a file")
	}
	if err := os.Remove(abs); err != nil {
		return false, err
	}
	return true, nil
}

// NormalizePath enforces that a workspace path is relative, free of
// `..` traversal and free of Windows-style absolute prefixes.
//
// Returns a cleaned forward-slash path on success.
func NormalizePath(p string) (string, error) {
	candidate := strings.TrimSpace(strings.ReplaceAll(p, `\`, `/`))
	if candidate == "" {
		return "", errors.New("workspace path is required")
	}
	if filepath.IsAbs(candidate) ||
		(len(candidate) >= 2 && candidate[1] == ':') /* C: */ {
		return "", errors.New("workspace paths must be relative")
	}
	parts := strings.Split(candidate, "/")
	var out []string
	for _, p := range parts {
		switch p {
		case "", ".":
			continue
		case "..":
			return "", errors.New("workspace paths cannot escape the notebook root")
		default:
			out = append(out, p)
		}
	}
	if len(out) == 0 {
		return "", errors.New("workspace path is required")
	}
	return strings.Join(out, "/"), nil
}

func loadFile(root, abs string, info os.FileInfo) (models.NotebookWorkspaceFile, error) {
	bytes, err := os.ReadFile(abs)
	if err != nil {
		return models.NotebookWorkspaceFile{}, err
	}
	rel, err := filepath.Rel(root, abs)
	if err != nil {
		return models.NotebookWorkspaceFile{}, err
	}
	mod := info.ModTime()
	if mod.IsZero() {
		mod = time.Now().UTC()
	}
	return models.NotebookWorkspaceFile{
		Path:      filepath.ToSlash(rel),
		Language:  inferLanguage(rel),
		Content:   string(bytes),
		SizeBytes: info.Size(),
		UpdatedAt: mod,
	}, nil
}

func inferLanguage(p string) string {
	ext := strings.ToLower(strings.TrimPrefix(filepath.Ext(p), "."))
	switch ext {
	case "py":
		return "python"
	case "sql":
		return "sql"
	case "r":
		return "r"
	case "md":
		return "markdown"
	case "json":
		return "json"
	case "ts", "tsx":
		return "typescript"
	case "js", "jsx":
		return "javascript"
	case "toml":
		return "toml"
	default:
		return "text"
	}
}
